### PR TITLE
rake generate:late_tasks

### PR DESCRIPTION
**Pull request by rgarner**
## [Generate CSV for late notification](https://trello.com/c/zkqPeEg9/868-generate-csv-for-late-notification)

For every incomplete task in a given year/month period window,
find all the users' emails in combination with all the frameworks that
are late and send them to a file that is Notify-friendly.

Generates to CSV to STDOUT by default (pipe to a file if
you know where it's going; no standard location is implied)

### CODE REVIEW

File: app/models/task/late_generator.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+          tasks.id,
+          frameworks.short_name AS framework_short_name,
+          users.name AS user_name,
+          users.email AS user_email
+        ')
+        .joins(
+          :framework,
+          supplier: { memberships: :user }
+        )
+    end
```
Comment (rgarner) [inline app/models/task/late_generator.rb]: I'm not sure this is doing what we want it to do exactly. There may well be validation errors on a submission for a task, but it doesn't mean that there isn't one `completed` somewhere else.

**Comment by tekin** [comment]
Also just to acknowledge this PR covers similar ground to https://github.com/dxw/DataSubmissionServiceAPI/pull/263, but that PR's existence was missed when this work was started. We will proceed with this PR because it includes the extra data that has been requested and addresses the issue with writing to a file.

**Comment by rgarner** [comment]
Good shout on the test data. I will glue more hair to the Data Yak™

**Comment by rgarner** [comment]
@tekin herring smell removed. Completed task added for flavour (pecorino, I think).

File: app/models/task/late_generator.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    def generate
+      logger.info "Generating late contacts for #{year}, #{month}"
+
+      output.puts(CSV.generate_line(HEADER))
+      user_framework_lates.each do |user_framework|
```
Comment (tekin) [inline app/models/task/late_generator.rb]: Approving but one last suggestion: perhaps for_each here to avoid future memory bloat?

**Pull request merged**